### ISSUE TITLE
Building on Arch Linux

### ISSUE THREAD

**Issue by xyproto** [fogleman/tracer#1]
Hi, when building with `make` I get:

```
Creating directories
Beginning release build
Compiling: src/main.cpp -> build/release/main.o
In file included from src/tracer/tracer.hpp:5,
                 from src/main.cpp:1:
src/tracer/embreespheres.hpp: In constructor ‘EmbreeSpheres::EmbreeSpheres(RTCDevice, const std::vector<EmbreeSphere>&, const std::vector<std::shared_ptr<Material> >&)’:
src/tracer/embreespheres.hpp:36:9: error: ‘memcpy’ was not declared in this scope
   36 |         memcpy(buf, spheres.data(), sizeof(EmbreeSphere) * m_NumSpheres);
      |         ^~~~~~
src/tracer/embreespheres.hpp:4:1: note: ‘memcpy’ is defined in header ‘<cstring ’; did you forget to ‘#include <cstring>’?
    3 | #include <embree3/rtcore.h>
  +++ |+#include <cstring>
    4 | #include <glm/glm.hpp>
make[1]: *** [Makefile:177: build/release/main.o] Error 1
make: *** [Makefile:117: release] Error 2
```

After adding `#include <cstring>` to `embreespheres.hpp` with:

```bash
sed -i '2s/^/#include <cstring>/' src/tracer/embreespheres.hpp
```

I get:

```
Creating directories
Beginning release build
Compiling: src/main.cpp -> build/release/main.o
Linking: bin/release/tracer
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../lib/libstdc++.so: undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
/usr/bin/ld: /usr/lib/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:167: bin/release/tracer] Error 1
make: *** [Makefile:117: release] Error 2
```

After adding a link flag to `-pthread` with:

```bash
sed -i 's/-lembree3/-lembree3 -pthread/g' Makefile
```

I get:

```
Creating directories
Beginning release build
Linking: bin/release/tracer
Making symlink: tracer -> bin/release/tracer
```

The finally when running `bin/release/tracer`, I get:

```
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_M_construct null not valid
zsh: abort (core dumped)  bin/release/tracer
```

I'm a fan of many of your Go projects, so I'm looking forward to trying this one out. Cheers! :slightly_smiling_face:

**Comment by fogleman** [comment]
I fixed the cstring and pthread issues.

The program expects a custom binary file as input, as mentioned in the readme this isn't quite ready for real usage yet.

**Comment by xyproto** [comment]
Thanks! I am aware that the project is under development, but I thought you might find it interesting nevertheless.

The latest `master` has none of the above issues, but this new one:

```
make
Creating directories
Beginning release build
Compiling: src/main.cpp -> build/release/main.o
In file included from /usr/include/c++/9.2.0/bits/random.h:35,
from /usr/include/c++/9.2.0/random:49,
from src/tracer/util.hpp:7,
from src/tracer/camera.hpp:7,
from src/tracer/tracer.hpp:4,
from src/main.cpp:1:
/usr/include/c++/9.2.0/bits/uniform_int_dist.h: In instantiation of ‘class std::uniform_int_distribution<double>’:
src/tracer/util.hpp:21:45:   required from here
/usr/include/c++/9.2.0/bits/uniform_int_dist.h:60:49: error: static assertion failed: template argument must be an integral type
60 |       static_assert(std::is_integral<_IntType>::value,
|                                                 ^~~~~
In file included from /usr/include/glm/detail/setup.hpp:690,
from /usr/include/glm/glm.hpp:105,
from src/tracer/box.hpp:3,
from src/tracer/tracer.hpp:3,
from src/main.cpp:1:
/usr/include/c++/9.2.0/type_traits: In instantiation of ‘struct std::make_unsigned<double>’:
/usr/include/c++/9.2.0/bits/uniform_int_dist.h:231:57:   required from ‘std::uniform_int_distribution<_IntType>::result_type std::uniform_int_distribution<_IntType>::operator()(_UniformRandomNumberGenerator&, const std::uniform_int_distribution<_IntType>::param_type&) [with_UniformRandomNumberGenerator = std::mersenne_twister_engine<long unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>; _IntType = double; std::uniform_int_distribution<_IntType>::result_type = double]’
/usr/include/c++/9.2.0/bits/uniform_int_dist.h:173:51:   required from ‘std::uniform_int_distribution<_IntType>::result_type std::uniform_int_distribution<_IntType>::operator()(_UniformRandomNumberGenerator&) [with _UniformRandomNumberGenerator = std::mersenne_twister_engine<long unsigned int, 32, 624, 397, 31, 2567483615, 11, 4294967295, 7, 2636928640, 15, 4022730752, 18, 1812433253>; _IntType = double; std::uniform_int_distribution<_IntType>::result_type = double]’
src/tracer/util.hpp:22:20:   required from here
/usr/include/c++/9.2.0/type_traits:1689:62: error: invalid use of incomplete type ‘class std::__make_unsigned_selector<double, false, false>’
1689 |     { typedef typename __make_unsigned_selector<_Tp>::__type type; };
|                                                              ^~~~
/usr/include/c++/9.2.0/type_traits:1596:11: note: declaration of ‘class std::__make_unsigned_selector<double, false, false>’
1596 |     class __make_unsigned_selector;
|           ^~~~~~~~~~~~~~~~~~~~~~~~
make[1]: *** [Makefile:177: build/release/main.o] Error 1
make: *** [Makefile:117: release] Error 2
```

I'll just close this issue, then, since filing issues is a bit premature. :)

**Comment by fogleman** [comment]
Good catch, not sure why that error doesn't show up on macOS. Fixed.
